tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -394,18 +394,126 @@ export class WebGPUBackend extends KernelBackend {
     return tensorData.values;
   }
 
-  // TODO: Remove once this is fixed:
-  // https://github.com/tensorflow/tfjs/issues/1595
   override readSync(dataId: object): BackendValues {
     const tensorData = this.tensorMap.get(dataId);
-    const {values} = tensorData;
+    const {values, complexTensorInfos} = tensorData;
 
-    if (values == null) {
-      throw new Error(
-          'WebGPU readSync is only available for CPU-resident tensors.');
+    if (values != null || tensorData.dtype === 'string') {
+      return values;
     }
 
-    return values;
+    if (tensorData.dtype === 'complex64') {
+      const realValues =
+          this.readSync(complexTensorInfos.real.dataId) as Float32Array;
+      const imagValues =
+          this.readSync(complexTensorInfos.imag.dataId) as Float32Array;
+      const complexVals = util.convertBackendValuesAndArrayBuffer(
+          backend_util.mergeRealAndImagArrays(realValues, imagValues).buffer,
+          'float32');
+      this.convertAndCacheOnCPU(dataId, complexVals);
+      return complexVals;
+    }
+
+    const alphaModes: GPUCanvasAlphaMode[] = ['opaque', 'premultiplied'];
+
+    const bufferInfo = tensorData.resourceInfo as BufferInfo;
+    const bufSize = bufferInfo.size;
+    util.assert(
+        bufSize % 4 === 0,
+        () => 'Because there is 4 bytes for' +
+            'one pixel, buffer size must be multiple of 4.');
```

Comment:
There should be a whitespace between `for` and `one` in the concatenated string.